### PR TITLE
Respect SNI with terminating gateways and log a warning if it isn't set alongside TLS

### PR DESCRIPTION
**Pull request by kyhavlov**
This PR adds a note to the terminating gateway docs strongly recommended using SNI when TLS is configured. It also adds a warning log message when configuring a service on a terminating gateway with TLS without SNI set.

### CODE REVIEW

File: agent/structs/config_entry.go [hunk 1]
```diff
@@ -82,6 +82,14 @@ type UpdatableConfigEntry interface {
 	ConfigEntry
 }
 
+// WarningConfigEntry is an optional interface implemented by a ConfigEntry
+// if it wants to be able to emit warnings when it is being upserted.
+type WarningConfigEntry interface {
+	Warnings() []string
```
Comment (kyhavlov) [inline agent/structs/config_entry.go:88]: I was thinking of making this part of the values returned by `Validate()` but that seemed like overkill for just this.

Comment (mkeeler) [reply agent/structs/config_entry.go:88]: I personally like the approach you took optional interface implementation and type assertions.

**Comment by freddygv** [comment]
This generally looks good. Is the SNI-based upstream SAN validation going to be in a separate PR?

**Review by mkeeler — COMMENTED** [review]
The changes here LGTM.

Could you also update the code snippets on [this page](https://www.consul.io/docs/connect/config-entries/terminating-gateway) to include the use of the SNI field.

**Comment by kyhavlov** [comment]
Added a test to include SNI being set - it only passes with the new changes.

**Review by rboyer — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/625370.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit 6cf22a5cefa98bb7e7c0e98ed4dd9eb59edfa762 onto `release/1.12.x` succeeded!

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 6cf22a5cefa98bb7e7c0e98ed4dd9eb59edfa762 onto `release/1.11.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/625370)

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 6cf22a5cefa98bb7e7c0e98ed4dd9eb59edfa762 onto `release/1.10.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/625370)

**Comment by hc-github-team-consul-core** [comment]
:cherries::x: Cherry pick of commit 6cf22a5cefa98bb7e7c0e98ed4dd9eb59edfa762 onto `release/1.9.x` failed! [Build Log](https://circleci.com/gh/hashicorp/consul/625370)

**Pull request merged**